src/typ.rs

```diff
@@ -1281,6 +1286,111 @@ impl<'a, 'b> Typer<'a, 'b> {
         }
     }
 
+    fn infer_bin(
+        &mut self,
+        elems: Vec<UntypedExprBinSegment>,
+        location: SrcSpan,
+    ) -> Result<TypedExpr, Error> {
+        let elems = elems
+            .into_iter()
+            .map(|s| self.infer_segment(*s.value, s.options, s.location))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Ok(TypedExpr::Bin {
+            location,
+            elems,
+            typ: bitstring(),
+        })
+    }
+
+    fn infer_segment(
+        &mut self,
+        value: UntypedExpr,
+        options: Vec<UntypedExprBinSegmentOption>,
+        location: SrcSpan,
+    ) -> Result<TypedExprBinSegment, Error> {
+        let value = self.infer(value)?;
+
+        let options = options
+            .into_iter()
+            .map(|option| self.infer_segment_option(option))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let type_specifier = BinaryTypeSpecifier::new(options.clone(), true)
```

Comment:
Could the BinaryTypeSpecifier hold a slice rather than cloning the vector? From the implementation I think it could borrow the options though I'm not 100% sure.

Reply:
I'm fairly sure this could just be a borrow.

Reply:
Indeed it could